vllm/core/scheduler.py

```diff
@@ -281,72 +302,214 @@ def _schedule(self) -> SchedulerOutputs:
+            if num_prompt_tokens > token_budget.get():
+                break
 
-        # Each sequence in the generation phase only takes one token slot.
-        # Therefore, the number of batched tokens is equal to the number of
-        # sequences in the RUNNING state.
-        num_batched_tokens = sum(
+            # The total number of sequences in the RUNNING state should not
+            # exceed the maximum number of sequences.
+            num_new_seqs = seq_group.get_max_num_running_seqs()
+            if (num_curr_seqs + num_new_seqs >
+                    self.scheduler_config.max_num_seqs):
+                break
+
+            if curr_loras is not None and lora_int_id > 0:
+                curr_loras.add(lora_int_id)
+            self.waiting.popleft()
+            self._allocate(seq_group)
+            prompting_seq_groups.append(seq_group)
+
+            token_budget.record_prefill_tokens(num_prompt_tokens)
+            num_curr_seqs += num_new_seqs
+            num_prompting_seqs += 1
+
+        self.waiting.extendleft(leftover_waiting_sequences)
+
+        return SchedulePrefillOutputs(token_budget, prompting_seq_groups,
+                                      ignored_seq_groups)
+
+    def _schedule(self) -> SchedulerOutputs:
+        token_budget = TokenBudget(
+            self.scheduler_config.max_num_batched_tokens)
+
+        # First schedule as many prefilling requests as possible,
+        # then schedule decoding requests.
+        num_curr_seqs = sum(
             seq_group.num_seqs(status=SequenceStatus.RUNNING)
             for seq_group in self.running)
+        prefilling_outputs = self._schedule_prefilling(token_budget,
+                                                       num_curr_seqs)
+
+        if len(prefilling_outputs.prompting_seq_groups) > 0:
+            decoding_outputs = SchedulerDecodeOutputs.create_empty()
+        else:
+            decoding_outputs = self._schedule_decoding(token_budget)
 
         scheduler_outputs = SchedulerOutputs(
-            scheduled_seq_groups=self.running,
-            prompt_run=False,
-            num_batched_tokens=num_batched_tokens,
-            blocks_to_swap_in=blocks_to_swap_in,
-            blocks_to_swap_out=blocks_to_swap_out,
-            blocks_to_copy=blocks_to_copy,
-            ignored_seq_groups=[],
+            scheduled_seq_groups=prefilling_outputs.prompting_seq_groups +
+            decoding_outputs.decoding_seq_groups,
+            num_prompt_groups=prefilling_outputs.num_selected_groups(),
+            num_batched_tokens=token_budget.num_batched_tokens,
+            blocks_to_swap_in=decoding_outputs.blocks_to_swap_in,
+            blocks_to_swap_out=decoding_outputs.blocks_to_swap_out,
+            blocks_to_copy=decoding_outputs.blocks_to_copy,
+            ignored_seq_groups=prefilling_outputs.ignored_seq_groups,
+            lora_enabled=self.lora_enabled,
         )
+
+        self.running.extend(prefilling_outputs.prompting_seq_groups)
```

Comment:
i like that the prefill/decode scheduling functions make it easier to implement both prefill and decode prioritizing schedules. but this seems incomplete, ideally the scheduling order should be:

1. Running decodes
2. Running prefills 
3. Swapped sequences - these can be either prefill and decode requests
4. New prefills

Reply:
This 4 steps are the order when chunked prefill is enabled actually (see https://github.com/vllm-project/vllm/pull/3483/files#diff-d3df23c3e3bcfe97ee8507061c6de54f0eff23a8c75d7f5999062c42245290f8). I will make sure swapped is properly separated to different methods and add tests! 

For this PR itself, I will preserve exact same ordering as previous PR. Let me write down the current ordering here and make sure the refactored logic contains the same order.

Reply:
Hmm actually I am reading the original code, and it seems like swap is only applied for decoding https://github.com/vllm-project/vllm/blob/e67c295b0c812a4969d5bbc70818f6d87c87f9e4/vllm/core/scheduler.py#L351. Basically when swap -> running state change happens, prompt_run is False.